File: solutions/04.enterprise-bot/Middleware/ContentModeratorMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Azure.CognitiveServices.ContentModerator;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.CognitiveServices.ContentModerator;

namespace EnterpriseBot
{
    /// <summary>
    /// Middleware component to run Content Moderator Service on all incoming activities.
    /// </summary>
    public class ContentModeratorMiddleware : IMiddleware
    {
        /// <summary>
        /// Key for Text Moderator result in Bot Context dictionary.
        /// </summary>
        public const string ServiceName = "ContentModerator";

        /// <summary>
        /// Key for Text Moderator result in Bot Context dictionary.
        /// </summary>
        public const string TextModeratorResultKey = "TextModeratorResult";

        /// <summary>
        /// Content Moderator service key.
        /// </summary>
        private readonly string subscriptionKey;

        /// <summary>
        /// Content Moderator service region.
        /// </summary>
        private readonly string region;

        /// <summary>
        /// Initializes a new instance of the <see cref="ContentModeratorMiddleware"/> class.
        /// </summary>
        /// <param name="subscriptionKey">Azure Service Key.</param>
        /// <param name="region">Azure Service Region.</param>
        public ContentModeratorMiddleware(string subscriptionKey, string region)
        {
            this.subscriptionKey = subscriptionKey;
            this.region = region;
        }

        /// <summary>
        /// Analyzes activity text with Content Moderator and adds result to Bot Context. Run on each turn of the conversation.
        /// </summary>
        /// <param name="context">The Bot Context object.</param>
        /// <param name="next">The next middleware component to run.</param>
        /// <param name="cancellationToken">The cancellation token for the task.</param>
        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
        public async Task OnTurnAsync(ITurnContext context, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            BotAssert.ContextNotNull(context);

            if (context.Activity.Type == ActivityTypes.Message)
            {
                var byteArray = Encoding.UTF8.GetBytes(context.Activity.Text);
                var textStream = new MemoryStream(byteArray);

                var client = new ContentModeratorClient(new ApiKeyServiceClientCredentials(this.subscriptionKey));
                client.BaseUrl = $"{this.region}.api.cognitive.microsoft.com";

                var screenResult = client.TextModeration.ScreenText(
                    language: "eng",
                    textContentType: "text/plain",
                    textContent: textStream,
                    autocorrect: true,
                    pII: true,
                    listId: null,
                    classify: true);

                context.TurnState.Add(TextModeratorResultKey, screenResult);
            }

            await next(cancellationToken);
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/middleware/contentModeratorMiddleware.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { ContentModeratorClient } from "azure-cognitiveservices-contentmoderator";
import { Screen } from "azure-cognitiveservices-contentmoderator/lib/models";
import { ActivityTypes, Middleware, TurnContext } from "botbuilder";
import { CognitiveServicesCredentials } from "ms-rest-azure";
import { Readable } from "stream";

/**
 * Middleware component to run Content Moderator Service on all incoming activities.
 */
export class ContentModeratorMiddleware implements Middleware {
    /**
     * Key for Text Moderator result in Bot Context dictionary.
     */
    public static readonly ServiceName: string = "ContentModerator";

    /**
     * Key for Text Moderator result in Bot Context dictionary.
     */
    public static readonly TextModeratorResultKey: string = "TextModeratorResult";

    private readonly _cmClient: ContentModeratorClient;

    /**
     * Initializes a new instance of the ContentModeratorMiddleware class.
     * @constructor
     * @param {string} subscriptionKey Azure Service Key.
     * @param {string} region Azure Service Region.
     */
    constructor(subscriptionKey: string, region: string) {
        this._cmClient = new ContentModeratorClient(new CognitiveServicesCredentials(subscriptionKey), `https://${region}.api.cognitive.microsoft.com`);
    }

    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
        if (context === null) {
            throw new Error("context is null");
        }

        if (context.activity.type === ActivityTypes.Message) {
            const content = new Readable();
            content.push(context.activity.text);
            content.push(null);
            const screenResult: Screen = await this._cmClient.textModeration.screenText("text/plain", content, { language: "eng", autocorrect: true, pII: true, classify: true });

            context.turnState.set(ContentModeratorMiddleware.TextModeratorResultKey, screenResult);
        }

        await next();
    }

}
